## Pandas I

Home exercises for Netflix:

1. Is there any missing rating?
2. How many films in 2021 correspond to your country?
3. What's the number of movies in 2020 with full information?
4. Give me the year with more titles,
5. and what has been the average in terms of releases from 2010. 

And for Titanic:

1. Calculate Gender-Based Survival Percentage

2. Calculate Survival Percentage Grouped by Gender and Class

### Netflix Dataset (df1)

In [23]:
import pandas as pd

path1 = '../session_6/netflix_titles.csv'

df1 = pd.read_csv(path1)
df1.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 8807 entries, 0 to 8806
Data columns (total 12 columns):
 #   Column        Non-Null Count  Dtype 
---  ------        --------------  ----- 
 0   show_id       8807 non-null   object
 1   type          8807 non-null   object
 2   title         8807 non-null   object
 3   director      6173 non-null   object
 4   cast          7982 non-null   object
 5   country       7976 non-null   object
 6   date_added    8797 non-null   object
 7   release_year  8807 non-null   int64 
 8   rating        8803 non-null   object
 9   duration      8804 non-null   object
 10  listed_in     8807 non-null   object
 11  description   8807 non-null   object
dtypes: int64(1), object(11)
memory usage: 825.8+ KB


#### Exercise 1: Is there any missing rating?

In [24]:
df1.isna().sum()

show_id            0
type               0
title              0
director        2634
cast             825
country          831
date_added        10
release_year       0
rating             4
duration           3
listed_in          0
description        0
dtype: int64

There are 4 missing ratings.

#### Exercise 2. How many films in 2021 correspond to your country?

In [25]:
df1['country'].value_counts()

country
United States                                                                                                                  2818
India                                                                                                                           972
United Kingdom                                                                                                                  419
Japan                                                                                                                           245
South Korea                                                                                                                     199
Canada                                                                                                                          181
Spain                                                                                                                           145
France                                                              

In [26]:
# Films from Singapore in 2021
films_sg_2021 = df1[(df1['country'] == 'Singapore') & (df1['release_year'] == 2021)]
films_sg_2021

Empty DataFrame
Columns: [show_id, type, title, director, cast, country, date_added, release_year, rating, duration, listed_in, description]
Index: []

There are 0 films from Singapore and in 2021.

#### Exercise 3: What's the number of movies in 2020 with full information?

In [ ]:
# Filter and display the rows that do not have missing values

df_clean = df1.dropna()
df_clean

show_id     type                          title             director  \
7         s8    Movie                        Sankofa         Haile Gerima   
8         s9  TV Show  The Great British Baking Show      Andy Devonshire   
9        s10    Movie                   The Starling       Theodore Melfi   
12       s13    Movie                   Je Suis Karl  Christian Schwochow   
24       s25    Movie                          Jeans           S. Shankar   
...      ...      ...                            ...                  ...   
8801   s8802    Movie                        Zinzana      Majid Al Ansari   
8802   s8803    Movie                         Zodiac        David Fincher   
8804   s8805    Movie                     Zombieland      Ruben Fleischer   
8805   s8806    Movie                           Zoom         Peter Hewitt   
8806   s8807    Movie                         Zubaan          Mozez Singh   

                                                   cast  \
7     Kofi Ghanaba, Oyafunmike Ogunlano, Alexandra D...   
8     Mel Giedroyc, Sue Perkins, Mary Berry, Paul Ho...   
9     Melissa McCarthy, Chris O'Dowd, Kevin Kline, T...   
12    Luna Wedler, Jannis Niewöhner, Milan Peschel, ...   
24    Prashanth, Aishwarya Rai Bachchan, Sri Lakshmi...   
...                                                 ...   
8801  Ali Suliman, Saleh Bakri, Yasa, Ali Al-Jabri, ...   
8802  Mark Ruffalo, Jake Gyllenhaal, Robert Downey J...   
8804  Jesse Eisenberg, Woody Harrelson, Emma Stone, ...   
8805  Tim Allen, Courteney Cox, Chevy Chase, Kate Ma...   
8806  Vicky Kaushal, Sarah-Jane Dias, Raaghav Chanan...   

                                                country          date_added  \
7     United States, Ghana, Burkina Faso, United Kin...  September 24, 2021   
8                                        United Kingdom  September 24, 2021   
9                                         United States  September 24, 2021   
12                              Germany, Czech Republic  September 23, 2021   
24                                                India  September 21, 2021   
...                                                 ...                 ...   
8801                       United Arab Emirates, Jordan       March 9, 2016   
8802                                      United States   November 20, 2019   
8804                                      United States    November 1, 2019   
8805                                      United States    January 11, 2020   
8806                                              India       March 2, 2019   

      release_year rating   duration  \
7             1993  TV-MA    125 min   
8             2021  TV-14  9 Seasons   
9             2021  PG-13    104 min   
12            2021  TV-MA    127 min   
24            1998  TV-14    166 min   
...            ...    ...        ...   
8801          2015  TV-MA     96 min   
8802          2007      R    158 min   
8804          2009      R     88 min   
8805          2006     PG     88 min   
8806          2015  TV-14    111 min   

                                             listed_in  \
7     Dramas, Independent Movies, International Movies   
8                         British TV Shows, Reality TV   
9                                     Comedies, Dramas   
12                        Dramas, International Movies   
24     Comedies, International Movies, Romantic Movies   
...                                                ...   
8801           Dramas, International Movies, Thrillers   
8802                    Cult Movies, Dramas, Thrillers   
8804                           Comedies, Horror Movies   
8805                Children & Family Movies, Comedies   
8806    Dramas, International Movies, Music & Musicals   

                                            description  
7     On a photo shoot in Ghana, an American model s...  
8     A talented batch of amateur bakers face off in...  
9     A woman adjusting to life after a loss contend...  
12    After 

In [28]:
movies_2020 = df_clean[df_clean['release_year'] == 2020]

num_movies_2020 = len(movies_2020)
num_movies_2020

442

There are 442 movies in 2020 with full information.

Exercise 4: Year with most titles

In [36]:
pd.set_option('display.max_rows', None) 

year_counts = df1['release_year'].value_counts()
year_counts = year_counts.sort_values(ascending=False)
year_counts.head(1)

release_year
2018    1147
Name: count, dtype: int64

The year with the most titles is 2018 with 1147 titles released.

#### Exercise 5: Average releases of titles from 2010. 

In [37]:
#number of titles from 2010 with information of release year:
num_titles_from_2010 =df1[df1['release_year'] >= 2010]['release_year'].count()
num_titles_from_2010

7472

In [38]:
from datetime import datetime

# Average number of titles released per year from 2010:

avg_titles_from_2010 = num_titles_from_2010 / (datetime.now().year - 2010)
avg_titles_from_2010

533.7142857142857

There is an average of 533 titles per year released from 2010.

### Titanic Dataset (df2)

In [39]:
path2 = '../session_6/train_and_test2.csv'

df2 = pd.read_csv(path2)
df2.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1309 entries, 0 to 1308
Data columns (total 28 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   Passengerid  1309 non-null   int64  
 1   Age          1309 non-null   float64
 2   Fare         1309 non-null   float64
 3   Sex          1309 non-null   int64  
 4   sibsp        1309 non-null   int64  
 5   zero         1309 non-null   int64  
 6   zero.1       1309 non-null   int64  
 7   zero.2       1309 non-null   int64  
 8   zero.3       1309 non-null   int64  
 9   zero.4       1309 non-null   int64  
 10  zero.5       1309 non-null   int64  
 11  zero.6       1309 non-null   int64  
 12  Parch        1309 non-null   int64  
 13  zero.7       1309 non-null   int64  
 14  zero.8       1309 non-null   int64  
 15  zero.9       1309 non-null   int64  
 16  zero.10      1309 non-null   int64  
 17  zero.11      1309 non-null   int64  
 18  zero.12      1309 non-null   int64  
 19  zero.1

In [41]:
df2.head()

Passengerid   Age     Fare  Sex  sibsp  zero  zero.1  zero.2  zero.3  \
0            1  22.0   7.2500    0      1     0       0       0       0   
1            2  38.0  71.2833    1      1     0       0       0       0   
2            3  26.0   7.9250    1      0     0       0       0       0   
3            4  35.0  53.1000    1      1     0       0       0       0   
4            5  35.0   8.0500    0      0     0       0       0       0   

   zero.4  ...  zero.12  zero.13  zero.14  Pclass  zero.15  zero.16  Embarked  \
0       0  ...        0        0        0       3        0        0       2.0   
1       0  ...        0        0        0       1        0        0       0.0   
2       0  ...        0        0        0       3        0        0       2.0   
3       0  ...        0        0        0       1        0        0       2.0   
4       0  ...        0        0        0       3        0        0       2.0   

   zero.17  zero.18  survived  
0        0        0         0  
1        0        0         1  
2        0        0         1  
3        0        0         1  
4        0        0         0  

[5 rows x 28 columns]

#### Exercise 1: Calculate Gender-Based Survival Percentage

In [42]:
# count of gender based on survival:
gender_survival = df2.groupby('Sex')['survived'].sum()
gender_survival

Sex
0    109
1    233
Name: survived, dtype: int64

In [43]:
total_pax_by_gender = df2['Sex'].value_counts()
total_pax_by_gender

Sex
0    843
1    466
Name: count, dtype: int64

- First, we create a Dataframe below that shows gender based survival and total passengers

- Then we concatenate 1 column that shows percentage of survival by gender

- From above we can see gender also has already been one-hot encoded into integers, looking at the lower survival absolute value of '0', we infer that '0' is Male and '1' is Female. I will then map the dataframe to an accurate gender naming as the index

In [44]:
gender_df = pd.DataFrame({
    'Gender Survival': gender_survival,
    'Total Passengers': total_pax_by_gender
})

In [45]:
gender_labels = {0: 'Male', 1: 'Female'}
gender_df.index = gender_df.index.map(gender_labels)
gender_df

Gender Survival  Total Passengers
Sex                                      
Male                109               843
Female              233               466

In [46]:
gender_df['Survival Percentage'] = ((gender_df['Gender Survival'] / gender_df['Total Passengers'])*100).round(2)

gender_df

Gender Survival  Total Passengers  Survival Percentage
Sex                                                           
Male                109               843                12.93
Female              233               466                50.00

#### Exercise 2: Calculate Survival Percentage Grouped by Gender and Class

In [47]:
#count of gender and class based on survival:
gender_class_survival = df2.groupby(['Sex', 'Pclass'])['survived'].sum()
gender_class_survival

Sex  Pclass
0    1         45
     2         17
     3         47
1    1         91
     2         70
     3         72
Name: survived, dtype: int64

Due to ambiguity in question in the base denominator to calculate a percentage, the base could either be:
(1) Total passengers in the Passenger Class with that Gender
(2) Total Overall Passengers in whole Titanic

Nevertheless, I will create a dataframe to illustrate both.

In [48]:
total_pax_gender_class = df2.groupby(['Sex', 'Pclass'])['survived'].count()
total_pax_gender_class


Sex  Pclass
0    1         179
     2         171
     3         493
1    1         144
     2         106
     3         216
Name: survived, dtype: int64

In [49]:
gender_class_df = pd.DataFrame({
    'Gender Class Survival': gender_class_survival,
    'Total Passengers': total_pax_gender_class
})
gender_class_df

Gender Class Survival  Total Passengers
Sex Pclass                                         
0   1                          45               179
    2                          17               171
    3                          47               493
1   1                          91               144
    2                          70               106
    3                          72               216

In [50]:
gender_class_df.index = gender_class_df.index.set_levels(gender_class_df.index.levels[0].map(gender_labels), level=0)

gender_class_df

Gender Class Survival  Total Passengers
Sex    Pclass                                         
Male   1                          45               179
       2                          17               171
       3                          47               493
Female 1                          91               144
       2                          70               106
       3                          72               216

In [51]:
gender_class_df['Survival Percentage By Gender Class'] = ((gender_class_df['Gender Class Survival'] / gender_class_df['Total Passengers'])*100).round(2)
gender_class_df['Overall Survival Percentage'] = ((gender_class_df['Gender Class Survival'] / len(df2))*100).round(2)

gender_class_df

Gender Class Survival  Total Passengers  \
Sex    Pclass                                            
Male   1                          45               179   
       2                          17               171   
       3                          47               493   
Female 1                          91               144   
       2                          70               106   
       3                          72               216   

               Survival Percentage By Gender Class  \
Sex    Pclass                                        
Male   1                                     25.14   
       2                                      9.94   
       3                                      9.53   
Female 1                                     63.19   
       2                                     66.04   
       3                                     33.33   

               Overall Survival Percentage  
Sex    Pclass                               
Male   1                              3.44  
       2                              1.30  
       3                              3.59  
Female 1                              6.95  
       2                              5.35  
       3                              5.50